src/libNetwork/RumorManager.cpp

```diff
@@ -0,0 +1,298 @@
+#include "RumorManager.h"
+
+#include <chrono>
+#include <thread>
+
+#include "P2PComm.h"
+#include "common/Messages.h"
+
+const unsigned char START_BYTE_GOSSIP = 0x33;
+
+namespace
+{
+    RRS::Message::Type convertType(uint8_t type)
+    {
+        switch (type)
+        {
+        case 1:
+            return RRS::Message::Type::PUSH;
+        case 2:
+            return RRS::Message::Type::PULL;
+        case 3:
+            return RRS::Message::Type::EMPTY_PUSH;
+        case 4:
+            return RRS::Message::Type::EMPTY_PULL;
+        default:
+            return RRS::Message::Type::UNDEFINED;
+        }
+    }
+
+} // anonymous namespace
+
+// CONSTRUCTORS
+RumorManager::RumorManager()
+    : m_peerIdPeerBimap()
+    , m_peerIdSet()
+    , m_rumorIdRumorBimap()
+    , m_selfPeer()
+    , m_rumorIdGenerator(0)
+    , m_mutex()
+    , m_continueRoundMutex()
+    , m_continueRound(false)
+    , m_condStopRound()
+{
+}
+
+RumorManager::~RumorManager() {}
+
+// PRIVATE METHODS
+
+void RumorManager::startRounds()
+{
+    LOG_MARKER();
+
+    std::thread([&]() {
+        std::unique_lock<std::mutex> guard(m_continueRoundMutex);
+        m_continueRound = true;
+        while (true)
+        {
+            { // critical section
+                std::lock_guard<std::mutex> guard(m_mutex);
+                std::pair<std::vector<int>, std::vector<RRS::Message>> result
+                    = m_rumorHolder->advanceRound();
+
+                // Get the corresponding Peer to which to send Push Messages if any.
+                for (const auto& i : result.first)
+                {
+                    auto l = m_peerIdPeerBimap.left.find(i);
+                    if (l != m_peerIdPeerBimap.left.end())
+                    {
+                        LOG_GENERAL(DEBUG,
+                                    "Sending " << result.second.size()
+                                               << " push messages");
+                        SendMessages(l->second, result.second);
+                    }
+                }
+            } // end critical section
+            if (m_condStopRound.wait_for(
```

Comment:
I'm curious if `stopRounds` is actually able to trigger `notify_all` because `m_continueRoundMutex` is held by this thread throughout its lifetime, and `stopRounds needs it`.  Were you able to check if this thread actually terminates when stopRounds is called?

Reply:
wait_for will actually release the m_continueRoundMutex for ROUND_TIME.